### PR TITLE
Feature test branch1

### PR DESCRIPTION
**Pull request by kodigantivandu**
Trying first pull request

### CODE REVIEW

File: Mars_Application_UserStory2_TestAutomation/Features/ExtendedPostive.feature [hunk 1]
```diff
@@ -0,0 +1,50 @@
+﻿Feature: ExtendedPostive
+Test MARS application Test cases
+
+Background: 
+## Positive_TC _001 - Log into Mars Application
```
Comment (amanbansalnz) [inline Mars_Application_UserStory2_TestAutomation/Features/ExtendedPostive.feature:5]: can you please break your feature files per feature not per type of test! feature file should be named Langauge.feature, skill.feature etc and within those you will need to put the types of tests extended postive extended negative etc

File: Mars_Application_UserStory2_TestAutomation/StepDefinitions/NegativeInvalidStepDefinitions.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+using Mars_Application_UserStory2_TestAutomation.Pages;
+using TechTalk.SpecFlow.Assist;
+using Assert = Xunit.Assert;
+
+namespace Mars_Application_UserStory2_TestAutomation.StepDefinitions
+{
+    [Binding]
+    public class NegativeInvalidStepDefinitions
+    {
+        private readonly IWebDriver driver;
```
Comment (amanbansalnz) [inline Mars_Application_UserStory2_TestAutomation/StepDefinitions/NegativeInvalidStepDefinitions.cs:17]: please make use of constructor! fix this every where you have not used a constructor see 
https://www.knowledgehut.com/tutorials/csharp/csharp-constructors

File: Mars_Application_UserStory2_TestAutomation/StepDefinitions/NegativeInvalidStepDefinitions.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        private readonly SignUpPage SignUpPageObj = new SignUpPage();
+        private readonly Login LoginPageObj = new Login();
+
+        public NegativeInvalidStepDefinitions(IWebDriver driver)
+        {
+            this.driver = driver;
```
Comment (amanbansalnz) [inline Mars_Application_UserStory2_TestAutomation/StepDefinitions/NegativeInvalidStepDefinitions.cs:23]: please create a common driver class which will initialise the driver, I see you created a hooks class which is good but a better approach would be to extend the Step and Page classes to Common driver class NegativeInvalidStepDefinitions : CommonDriver, In the Hook Class create the instance of Common driver and then class the initialise method for example 

   public class CommonDriver
    {
        public IWebDriver Initialise()
        {

            ChromeOptions options = new ChromeOptions();

            // Set window size to 90% of the screen size
            options.AddArgument("force-device-scale-factor=1");

            return new ChromeDriver(options);
        }

        public static void CloseDriver(IWebDriver driver1)
        {
            if (driver1 != null)
            {
                driver1.Quit();
                driver1 = null;
            }
        }
    }
 and in the Hooks class



public class SpecFlowHooks
{
    private static CommonDriver commonDriver;

    [BeforeTestRun] 
    public static void BeforeTestRun()
    {
        commonDriver = new CommonDriver();
        commonDriver.Initialise();
    }
}

File: Mars_Application_UserStory2_TestAutomation/Pages/ProfilePage.cs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+using Assert = Xunit.Assert;
+
+namespace Mars_Application_UserStory2_TestAutomation.Pages
+{
+    public class ProfilePage
+    {
+        public void CreateLanguage(IWebDriver driver, string language, string languageLevel)
+        {
+            Thread.Sleep(1000);
+            IWebElement AddnewButton = driver.FindElement(By.XPath("//*[@id=\"account-profile-section\"]/div/section[2]/div/div/div/div[3]/form/div[2]/div/div[2]/div/table/thead/tr/th[3]/div"));
```
Comment (amanbansalnz) [inline Mars_Application_UserStory2_TestAutomation/Pages/ProfilePage.cs:20]: you need to implement the POM pattern please https://docs.specflow.org/projects/specflow/en/latest/ui-automation/Selenium-with-Page-Object-Pattern.html 
also fix this for all pages

File: Mars_Application_UserStory2_TestAutomation/Pages/ProfilePage.cs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+            AddnewButton.Click();
+            Thread.Sleep(2000);
+
+            IWebElement AddLanguage = driver.FindElement(By.Name("name"));
+            AddLanguage.SendKeys(language);
+            Thread.Sleep(1000);
```
Comment (amanbansalnz) [inline Mars_Application_UserStory2_TestAutomation/Pages/ProfilePage.cs:26]: please use waits

File: Mars_Application_UserStory2_TestAutomation/Pages/ProfilePage.cs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+using Assert = Xunit.Assert;
+
+namespace Mars_Application_UserStory2_TestAutomation.Pages
+{
+    public class ProfilePage
+    {
+        public void CreateLanguage(IWebDriver driver, string language, string languageLevel)
```
Comment (amanbansalnz) [inline Mars_Application_UserStory2_TestAutomation/Pages/ProfilePage.cs:17]: you need to thing about state 

example before adding language you need to make sure there a no existing languages as this could result in false positive test pass
you will need to think about this for all the scenarios

the state is say that you have a test that is adding a language english, the second time you run this test you should delete any existing language so that the state is reset, this will guarantee that your test does not produce false negatives and passes always
you can assume that test execution order is not always guaranteed
imagine if you run the tests  in parallel mode the all tests will run at the same time
so you need to also make sure you do not wipe out other tests data
hence why when you remove data the best thing to do is delete only data you added for the test to be correct
e.g test 1 added english and french as language test 2 added hindi then when you reset data you already know what language you are adding for each test
there for on reset of the two test you should delete english and french for test 1 and hindi for test 2

Comment (amanbansalnz) [inline Mars_Application_UserStory2_TestAutomation/Pages/ProfilePage.cs:17]: please do not pass the driver around the action methods 

CreateLanguage(IWebDriver driver, string language, string languageLevel)

should be 

CreateLanguage(string language, string languageLevel)

you will either need to extend to Common driver class or you will need to pass driver via the constructor 

private IWebDriver driver,
ProfilePage(IWebDriver driver){

this.driver =driver;
}
